### PR TITLE
[SPARK-20994] Remove redundant characters in OpenBlocks to save memory for shuffle service.

### PR DESCRIPTION
**Pull request by jinxing64**
## What changes were proposed in this pull request?

In current code, blockIds in `OpenBlocks` are stored in the iterator on shuffle service. 
There are some redundant characters in  blockId(`"shuffle_" + shuffleId + "_" + mapId + "_" + reduceId`). This pr proposes to improve the footprint and alleviate the memory pressure on shuffle service.

### CODE REVIEW

**Comment by jinxing64** [comment]
n my cluster, we are suffering from OOM of shuffle-service.
We found that a lot of executors are fetching blocks from a single shuffle-service. Analyzing the memory, we found that the blockIds(shuffle_shuffleId_mapId_reduceId) takes about 1.5GBytes.

**Review by srowen — COMMENTED** [review]
Can that really save much memory? Seems trivial

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 1]
```diff
@@ -209,4 +190,47 @@ private ShuffleMetrics() {
...
+      if (blockIdParts.length < 4) {
+        throw new IllegalArgumentException("Unexpected block id format: " + blockId);
+      } else if (!blockIdParts[0].equals("shuffle")) {
+        throw new IllegalArgumentException("Expected shuffle block id, got: " + blockId);
+      }
+      this.shuffleId = blockIdParts[1];
+      mapIdAndReduceIds = new byte[blockIds.length][];
+      if (blockIds.length > 0) {
+        for (int i = 0; i< blockIds.length; i++) {
+          mapIdAndReduceIds[i] = (blockIdParts[2] + "_" + blockIdParts[3]).getBytes();
```
Comment (vanzin) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: Instead of storing this as a byte array, how about storing them as ints or longs (depending on what's the actual data type of the id)?

e.g., instead of:

```
private byte[][] mapIdAndReduceIds;
```

Which results in `blockIds.length + 1` arrays in total, you could have a single one where for each block id you have two entries, one for map id and one for reduce id, or something along those lines.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 2]
```diff
@@ -209,4 +190,47 @@ private ShuffleMetrics() {
...
+
+    private int index = 0;
+    private String appId;
+    private String execId;
+    private String shuffleId;
+    // mapId and reduceId are stored as bytes to save memory.
+    private byte[][] mapIdAndReduceIds;
+
+    ManagedBufferIterator(String appId, String execId, String[] blockIds) {
+      this.appId = appId;
```
Comment (vanzin) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: Wonder if you see a lot of these in your heap dump too? You could potentially intern `appId` and `execId` for some extra memory savings, if you see a lot of those.

**Comment by jinxing64** [comment]
@srowen 
Thanks a lot looking into this :)
For example: blockId="shuffle_20_1000_2000", it is stored as an `String`, which costs more than 20 bytes. In this change, it will cost only 8 bytes.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java (hunk 2 above)
Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: @vanzin 
There's one `appId` and `execId` per stream. I don't see a lot in my heap dump. Do you have any thoughts for interning this? :)

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java (hunk 1 above)
Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: Yes, I think this is a good idea. In current change, I make it to be `int[blockIds.length][2]`. I'm not sure if I understand your comment correctly. Please take another look :)

**Comment by jinxing64** [comment]
@vanzin 
Thanks a lot for reviewing this. I refined according to your comments, Please take another look at this when you have time :)

**Comment by srowen** [comment]
That's 12 bytes. Are there millions of these?

**Comment by jinxing64** [comment]
Actually it's more than 12 bytes. 
Yes, there are millions of these. In my heap dump, it's 1.5 G

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 3]
```diff
@@ -209,4 +190,52 @@ private ShuffleMetrics() {
...
+      this.execId = execId;
+      String[] blockId0Parts = blockIds[0].split("_");
+      if (blockId0Parts.length < 4) {
+        throw new IllegalArgumentException("Unexpected block id format: " + blockIds[0]);
+      } else if (!blockId0Parts[0].equals("shuffle")) {
+        throw new IllegalArgumentException("Expected shuffle block id, got: " + blockIds[0]);
+      }
+      this.shuffleId = blockId0Parts[1];
+      mapIdAndReduceIds = new int[blockIds.length][2];
+      if (blockIds.length > 0) {
```
Comment (srowen) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: This is superfluous

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 4]
```diff
@@ -209,4 +190,52 @@ private ShuffleMetrics() {
...
+    // An array containing mapId and reduceId pairs.
+    private int[][] mapIdAndReduceIds;
+
+    ManagedBufferIterator(String appId, String execId, String[] blockIds) {
+      this.appId = appId;
+      this.execId = execId;
+      String[] blockId0Parts = blockIds[0].split("_");
+      if (blockId0Parts.length < 4) {
+        throw new IllegalArgumentException("Unexpected block id format: " + blockIds[0]);
+      } else if (!blockId0Parts[0].equals("shuffle")) {
```
Comment (srowen) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: You don't need the 'else' here

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 5]
```diff
@@ -209,4 +190,52 @@ private ShuffleMetrics() {
     }
   }
 
+  private class ManagedBufferIterator implements Iterator<ManagedBuffer> {
```
Comment (srowen) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: Why break this out -- it's not necessary for the change right? just for clarity?

**Review by srowen — CHANGES_REQUESTED** [review]
Pardon, I'm missing how this saves memory somewhere -- where is a string stored that's now a shorter string?

**Comment by jinxing64** [comment]
@srowen  Sorry, I didn't make it clear.
1. In current code, all blockIds are stored in the iterator. They are released only when the iterator is traversed.
2. Now I change the `String` to be two `int`

**Comment by srowen** [comment]
The current iterator doesn't have any state except for an int. What are you referring to?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java (hunk 5 above)
Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: I think the iterator is becoming a little bit complicated. So I break this out and give a constructor.

**Comment by jinxing64** [comment]
I mean the blockIds in `OpenBlocks`, they have reference in iterator.

**Comment by srowen** [comment]
I get it. But that doesn't make the reference in OpenBlocks go away. This only helps anything is msg/msgObj can be garbage collected earlier. Is that the case? right now this is allocating additional memory, not instead of the existing memory.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java (hunk 4 above)
Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: We have some kinds of `BlockId`, I guess it's better to have a check here and we can parse the blockId correctly.

**Comment by jinxing64** [comment]
The blockIds cannot be freed because they are referenced in the iterator. In current change they are not. We reference the mapIdAndReduceIds instead. Thus the blockIds in OpenBlocks can be garbage collected.

**Comment by srowen** [comment]
That's not the question though. The question is whether they could be freed even after this change. msg still references it. That's what you need to establish, if only by some empirical testing.

**Comment by jinxing64** [comment]
there is no where referencing `msg`, right? I guess the `msg` will be garbage collected fluently.

**Comment by srowen** [comment]
I'm not clear that's true, no. Not, at least, in the lifetime of the iterator. That's what has to be true for this to help anything. Do you have evidence this is true? for example if you have tests that clearly show the memory is released earlier, that would be good evidence.

**Comment by jinxing64** [comment]
Yes, I think it's great to do some tests and give a good evidence.

**Comment by jinxing64** [comment]
@srowen 
I did a test to verify this patch. 
I wrap a number of blocks inside `OpenBlocks` and send it to `ExternalShuffleBlockHandler`.
With this change:
it cost about 133M in the memory; analyzing heap dump, there is only int[][], `blockIds` is released. 
Without this change,:
it cost about 362M in the memory; analyzing heap dump, there is String[].

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 6]
```diff
@@ -209,4 +190,51 @@ private ShuffleMetrics() {
...
+
+    private int index = 0;
+    private String appId;
+    private String execId;
+    private String shuffleId;
+    // An array containing mapId and reduceId pairs.
+    private int[][] mapIdAndReduceIds;
```
Comment (vanzin) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: Actually, I mean a single array. e.g.

```
int[] mapIdAndReduceIds;

mapIdAndReduceIds = new int[blockIds.length * 2];
mapIdAndReduceIds[0] = mapId1;
mapIdAndReduceIds[1] = reduceId1;
mapIdAndReduceIds[2] = mapId2;
mapIdAndReduceIds[3] = reduceId2;
etc etc etc
```

Reason being that if you really have millions of these, each "child" array in your two-dimensional array wastes 16 (or 20?) bytes (16 bytes of object overhead + 4 bytes for the array length). Looking in jvisual vm, an empty array actually consumes 24 bytes, so it seems the JVM is aligning things and wasting an extra 4 bytes per array...

**Comment by vanzin** [comment]
@srowen I don't see any references to the original `OpenBlocks` message nor to the block id array in the updated code, not sure why do you think there's still a reference somewhere?

**Comment by srowen** [comment]
There isn't a reference _here_ anymore; there could be elsewhere. It sounds like there's good reason to believe there is not another reference hanging around though.

**Comment by vanzin** [comment]
> There isn't a reference here anymore; there could be elsewhere.

Only if there was a bug in the RPC layer, since this is an RPC handler and the message should not be referenced by the RPC code after the method returns.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 7]
```diff
@@ -209,4 +190,51 @@ private ShuffleMetrics() {
     }
   }
 
+  private class ManagedBufferIterator implements Iterator<ManagedBuffer> {
+
+    private int index = 0;
+    private String appId;
```
Comment (vanzin) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: `final`, same for all fields except `index`.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java (hunk 4 above)
Comment (vanzin) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: I think Sean means that since you're throwing in the previous block, `else` is redundant.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 8]
```diff
@@ -209,4 +190,51 @@ private ShuffleMetrics() {
...
+    @Override
+    public boolean hasNext() {
+      return index < mapIdAndReduceIds.length / 2;
+    }
+
+    @Override
+    public ManagedBuffer next() {
+      String blockId = "shuffle_" + shuffleId + "_" + mapIdAndReduceIds[2 * index] + "_" +
+        mapIdAndReduceIds[2 * index + 1];
+      final ManagedBuffer block = blockManager.getBlockData(appId, execId, blockId);
```
Comment (vanzin) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: Is it too big a change to make `getBlockData` expect the broken down shuffle/map/reduce id, since all it does is parse the string again into integers?

**Review by vanzin — COMMENTED** [review]
Is this still a "WIP"?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java (hunk 7 above)
Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: Yes, I will refine.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java (hunk 8 above)
Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: I was thinking about this. I added a new `getBlockData` in ExternalShuffleBlockResolver`. But not sure if it is appropriate.

**Comment by jinxing64** [comment]
@vanzin 
Thanks again for comments :) I refined accordingly, please take another look when you have time.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java [hunk 9]
```diff
@@ -164,7 +163,15 @@ public ManagedBuffer getBlockData(String appId, String execId, String blockId) {
...
     int mapId = Integer.parseInt(blockIdParts[2]);
     int reduceId = Integer.parseInt(blockIdParts[3]);
+    return getBlockData(appId, execId, shuffleId, mapId, reduceId);
+  }
 
+  /**
+   * Obtains a FileSegmentManagedBuffer from (shuffleId, mapId, reduceId). We make assumptions
+   * about how the hash and sort based shuffles store their data.
+   */
+  public ManagedBuffer getBlockData(String appId, String execId, int shuffleId, int mapId,
```
Comment (vanzin) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java]: nit: style. See constructor at top of file for the style when param lists are long.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java [hunk 10]
```diff
@@ -151,8 +151,7 @@ public void registerExecutor(
 
   /**
    * Obtains a FileSegmentManagedBuffer from a shuffle block id. We expect the blockId has the
-   * format "shuffle_ShuffleId_MapId_ReduceId" (from ShuffleBlockId), and additionally make
-   * assumptions about how the hash and sort based shuffles store their data.
+   * format "shuffle_ShuffleId_MapId_ReduceId" (from ShuffleBlockId).
    */
   public ManagedBuffer getBlockData(String appId, String execId, String blockId) {
```
Comment (vanzin) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java]: Is this method used anywhere else? I only see `ExternalShuffleBlockHandler` using this class, and it now uses the new method. If only unit tests use this, then remove this method and fix the unit tests.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 11]
```diff
@@ -209,4 +190,51 @@ private ShuffleMetrics() {
...
+      String[] blockId0Parts = blockIds[0].split("_");
+      if (blockId0Parts.length < 4) {
+        throw new IllegalArgumentException("Unexpected block id format: " + blockIds[0]);
+      }
+      if (!blockId0Parts[0].equals("shuffle")) {
+        throw new IllegalArgumentException("Expected shuffle block id, got: " + blockIds[0]);
+      }
+      this.shuffleId = Integer.parseInt(blockId0Parts[1]);
+      mapIdAndReduceIds = new int[2 * blockIds.length];
+      for (int i = 0; i< blockIds.length; i++) {
```
Comment (vanzin) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: nit: space before `<`

**Review by vanzin — COMMENTED** [review]
One nit otherwise LGTM. I'll leave it overnight in case others want to take a look.

**Comment by vanzin** [comment]
(also PR title has a typo, should be "redundant")

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 12]
```diff
@@ -209,4 +190,51 @@ private ShuffleMetrics() {
...
+        throw new IllegalArgumentException("Unexpected block id format: " + blockIds[0]);
+      }
+      if (!blockId0Parts[0].equals("shuffle")) {
+        throw new IllegalArgumentException("Expected shuffle block id, got: " + blockIds[0]);
+      }
+      this.shuffleId = Integer.parseInt(blockId0Parts[1]);
+      mapIdAndReduceIds = new int[2 * blockIds.length];
+      for (int i = 0; i < blockIds.length; i++) {
+        String[] blockIdParts = blockIds[i].split("_");
+        if (Integer.parseInt(blockIdParts[1]) != shuffleId) {
```
Comment (jiangxb1987) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: blockIdParts[1].toInt != shuffleId ?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 13]
```diff
@@ -209,4 +190,51 @@ private ShuffleMetrics() {
...
+      String[] blockId0Parts = blockIds[0].split("_");
+      if (blockId0Parts.length < 4) {
+        throw new IllegalArgumentException("Unexpected block id format: " + blockIds[0]);
+      }
+      if (!blockId0Parts[0].equals("shuffle")) {
+        throw new IllegalArgumentException("Expected shuffle block id, got: " + blockIds[0]);
+      }
+      this.shuffleId = Integer.parseInt(blockId0Parts[1]);
+      mapIdAndReduceIds = new int[2 * blockIds.length];
+      for (int i = 0; i < blockIds.length; i++) {
```
Comment (jiangxb1987) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: How about rewrite this to be imperative?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java [hunk 14]
```diff
@@ -150,27 +150,20 @@ public void registerExecutor(
...
+      String execId,
+      int shuffleId,
+      int mapId,
+      int reduceId) {
     ExecutorShuffleInfo executor = executors.get(new AppExecId(appId, execId));
     if (executor == null) {
       throw new RuntimeException(
         String.format("Executor is not registered (appId=%s, execId=%s)", appId, execId));
     }
-
```
Comment (jiangxb1987) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java]: nit: we should keep the original format.

File: common/network-shuffle/src/test/java/org/apache/spark/network/sasl/SaslIntegrationSuite.java [hunk 15]
```diff
@@ -202,7 +202,7 @@ public void onBlockFetchFailure(String blockId, Throwable t) {
         }
       };
 
-      String[] blockIds = { "shuffle_2_3_4", "shuffle_6_7_8" };
+      String[] blockIds = { "shuffle_0_1_2", "shuffle_0_3_4" };
```
Comment (jiangxb1987) [inline common/network-shuffle/src/test/java/org/apache/spark/network/sasl/SaslIntegrationSuite.java]: What's the purpose of this change?

**Review by jiangxb1987 — COMMENTED** [review]
The change seems reasonable, but the code style still need to be improved. Also cc @cloud-fan to make a pass.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 16]
```diff
@@ -209,4 +190,51 @@ private ShuffleMetrics() {
...
+    private final String execId;
+    private final int shuffleId;
+    // An array containing mapId and reduceId pairs.
+    private final int[] mapIdAndReduceIds;
+
+    ManagedBufferIterator(String appId, String execId, String[] blockIds) {
+      this.appId = appId;
+      this.execId = execId;
+      String[] blockId0Parts = blockIds[0].split("_");
+      if (blockId0Parts.length < 4) {
```
Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: I was thinking to throw the `IllegalArgumentException`.
Pardon, I'm not sure how to use `require` in java.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java (hunk 12 above)
Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: It's hard to do that in java, right?

File: common/network-shuffle/src/test/java/org/apache/spark/network/sasl/SaslIntegrationSuite.java (hunk 15 above)
Comment (jinxing64) [inline common/network-shuffle/src/test/java/org/apache/spark/network/sasl/SaslIntegrationSuite.java]: With this change ,we cannot shuffle blocks with multiple `shuffleIds`

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java (hunk 16 above)
Comment (jiangxb1987) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: nvm, didn't notice they are java code.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java (hunk 13 above)
Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: Pardon, could you give an example?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java (hunk 14 above)
Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java]: There are two blank lines originally. I guess it's appropriate to remove one?

**Comment by jinxing64** [comment]
@jiangxb1987 
Thanks a lot for taking time review this pr. More comments are welcome.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java (hunk 16 above)
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: shall we be more strict and use `blockId0Parts.length != 4`?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 17]
```diff
@@ -209,4 +190,51 @@ private ShuffleMetrics() {
...
+        throw new IllegalArgumentException("Unexpected block id format: " + blockIds[0]);
+      }
+      if (!blockId0Parts[0].equals("shuffle")) {
+        throw new IllegalArgumentException("Expected shuffle block id, got: " + blockIds[0]);
+      }
+      this.shuffleId = Integer.parseInt(blockId0Parts[1]);
+      mapIdAndReduceIds = new int[2 * blockIds.length];
+      for (int i = 0; i < blockIds.length; i++) {
+        String[] blockIdParts = blockIds[i].split("_");
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: shall we check `blockIdParts[0] == "shufle"`?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 18]
```diff
@@ -209,4 +190,51 @@ private ShuffleMetrics() {
...
+            ", got:" + blockIds[i]);
+        }
+        mapIdAndReduceIds[2 * i] = Integer.parseInt(blockIdParts[2]);
+        mapIdAndReduceIds[2 * i + 1] = Integer.parseInt(blockIdParts[3]);
+      }
+    }
+
+    @Override
+    public boolean hasNext() {
+      return index < mapIdAndReduceIds.length / 2;
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: nit: we can keep a `pos`, and increase it by 2 in `next`, so here we can just write `pos < mapIdAndReduceIds.length` to save a division.

**Comment by cloud-fan** [comment]
LGTM except some minor comments

**Comment by jinxing64** [comment]
@cloud-fan 
Thanks a lot for taking time review this. I refined accordingly :)

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java [hunk 19]
```diff
@@ -209,4 +190,51 @@ private ShuffleMetrics() {
...
+    private final String execId;
+    private final int shuffleId;
+    // An array containing mapId and reduceId pairs.
+    private final int[] mapIdAndReduceIds;
+
+    ManagedBufferIterator(String appId, String execId, String[] blockIds) {
+      this.appId = appId;
+      this.execId = execId;
+      String[] blockId0Parts = blockIds[0].split("_");
+      if (blockId0Parts.length < 4 || !blockId0Parts[0].equals("shuffle")) {
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: use `blockId0Parts.length != 4`?

Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java]: Sure.

**Comment by jinxing64** [comment]
Jenkins, retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by jinxing64** [comment]
@cloud–fan
Thanks for merging !
